hibernate-core/src/main/java/org/hibernate/engine/query/spi/QueryPlanCache.java

```diff
@@ -152,7 +153,7 @@ public HQLQueryPlan getHQLQueryPlan(String queryString, boolean shallow, Map<Str
 			final long startTime = ( stats ) ? System.nanoTime() : 0L;
 
 			LOG.tracev( "Unable to locate HQL query plan in cache; generating ({0})", queryString );
-			value = new HQLQueryPlan( queryString, shallow, enabledFilters, factory );
+			value = createHQLQueryPlan( queryString, shallow, enabledFilters, factory );
```

Comment:
Are you guys really using inheritance everywhere?  You *really* want to subclass all of these?  Why not just pass in functionals that can create the plan?  Then your SessionFactory (I assume you already subclass SessionFactory, Session, etc) can simply pass in a different one.

I mean, feel free to keep subclassing everything, but personally I'd be trying to minimize what I need to subclass

Reply:
@sebersole the new subclass `RxHQLQueryPlan` actually adds new public operations, it's not just overriding operations of `HQLQueryPlan`.

On the other hand, perhaps you're right about `RxQueryPlanCache`: I now notice we could probably eliminate that class and replace it by a new argument to `QueryPlanCache`. @DavideD do you want me to try and make that change?

Reply:
I think I can take care of it. Let me give it a go.

Reply:
I've almost got it done.

Reply:
@DavideD can I push to your branch?

Reply:
You are talking about adding a function for the creation of the QueryPlanCache, isn't it? Can I add a new `getHQLQueryPlan` with this additional argument? Or do  need to keep the API of `QueryPlanCache` the same?

Reply:
OK, push it on the branch. No problem

Reply:
@DavideD do I have permission?

    Authentication failed for 'https://github.com/DavideD/hibernate-orm.git/'

Reply:
Anyway the commit is c0ca1560e8bf385e0c3b9431b118780a8e0953e0

Reply:
And the corresponding change to your patch to hibernate-reactive is https://github.com/gavinking/hibernate-rx/commit/2f3d48ff4514225219d2855b3c988e778e11e327

Reply:
I've updated both PRs. I've squashed the commits in the Hibernate Reactive one